Fix base64 implementation in pmix framework.

artpol84 opened the pull request:
In the commit 80f07b65f16e9538aca7fc5e124d2074e7e0b69e setting of '-' marker used as the string termination sign was moved from base64 code:
from: https://github.com/open-mpi/ompi/commit/80f07b65f16e9538aca7fc5e124d2074e7e0b69e#diff-1b10896c267d2591dc2c08fd0542ab67L491
to: https://github.com/open-mpi/ompi/commit/80f07b65f16e9538aca7fc5e124d2074e7e0b69e#diff-1b10896c267d2591dc2c08fd0542ab67R189

However the decoding function wasn't fixed and still expects on extra byte at the end of the encoded string.
This leads to **data truncation during decoding** (was noticed on standalone code that was using base64 from OMPI).

artpol84 commented:
This needs to go to v2.x as well.

artpol84 commented:
@ggouaillardet @rhc54 @hppritcha please review.

artpol84 commented:
:bot:retest

ggouaillardet commented:
that does not look quite right to me
if I understand correctly, the base64 input (aka data) is "- " terminated
the length of the binary result data is
(strlen(data)-1)*3/4
since these are integer, multiplication should be performed first

for example, if you encode 1 byte, the encoded string will be "xy-"
with this commit, 0 bytes are calloc'ed

makes sense ?

artpol84 commented:
@ggouaillardet, disagree. You are talking about encoding logic. And this PR addresses **decoding** logic. The formula you referring wasn't changed here.
If you have `n` bytes to encode you will get `m = ceil (n / 3) * 4` encoded bytes. `n % 3 <> 0` you will have some empty bytes (equal to 64) at the end of the encoded sequence. But in any case it will be exhaustive.

In current code I was encoding 4 bytes and getting only 3 of them back:
original: 4 input bytes = 3 + 1 = > encoded 4 + 4 = 8 bytes.

With current formula you'll have  strlen() = 8 => (8  - 1) / 4 = 7 / 4 = 1 - one block to unpack instead of 2!

ggouaillardet commented:
so the string to be decoded (aka data) is no more "-" terminated ?
I am surprised 1 byte is not encoded into 2 ascii chars (3 if you add the "-" terminator) but 4.
anyway, I will review the encoding/decoding logic more thoroughly tomorrow

artpol84 commented:
1. I think that strings are still '-' terminated but not by base64 code anymore, so this symbol is accounted in the plain data length already.
2. As far as I understood the code each 3 chars of the plain data are encoded with 4 characters of base64 data. And if you have 4 bytes you will end up having 2 4-byte blocks. The first will be full and contain first 3 bytes and the second one (tail) will not be full and represent only one byte. It is fine since this only happens with the last base64 block.

ggouaillardet commented:
ok, the encoded string length is a multiple of 4 (+1 if "-" terminated)
in pmix, pmi_decode null terminates the result, which makes no sense since the result is binary
(on top of that, this causes an overflow...)

artpol84 commented:
Indeed we don't need to null terminate the binary.
But what do you mean by overflow. As far as I was able to see during
encoding we account null byte. And during decoding this doesn't have any
effect.

понедельник, 23 мая 2016 г. пользователь Gilles Gouaillardet написал:

> ok, the encoded string length is a multiple of 4 (+1 if "-" terminated)
> in pmix, pmi_decode null terminates the result, which makes no sense since
> the result is binary
> (on top of that, this causes an overflow...)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/open-mpi/ompi/pull/1688#issuecomment-220970009>
>


-- 
-----
Best regards, Artem Polyakov
(Mobile mail)

ggouaillardet commented:
if we null terminate the binary, then we need to calloc +1 byte.
if the to be encoded binary is a null terminated string, then we have to encode the string and the null terminator, so a null terminated string will be decoded
in any case, we cannot null terminate in pmi_decode unless an extra byte is calloc'ed

artpol84 commented:
I recalled why we null terminate the binary. Encoded string goes to SLURM
PMI which assumes a string. So encoded string has to be null terminated.

понедельник, 23 мая 2016 г. пользователь Gilles Gouaillardet написал:

> if we null terminate the binary, then we need to calloc +1 byte.
> if the to be encoded binary is a null terminated string, then we have to
> encode the string and the null terminator, so a null terminated string will
> be decoded
> in any case, we cannot null terminate in pmi_decode unless an extra byte
> is calloc'ed
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/open-mpi/ompi/pull/1688#issuecomment-220982156>
>


-- 
-----
Best regards, Artem Polyakov
(Mobile mail)

artpol84 commented:
As for encoding null terminated string - I don't see any issues with this -
you just need to pass the right length to encode, base64 will handle all
null chars. In fact current OMPI meta keys have a lot of nulls.

понедельник, 23 мая 2016 г. пользователь Gilles Gouaillardet написал:

> if we null terminate the binary, then we need to calloc +1 byte.
> if the to be encoded binary is a null terminated string, then we have to
> encode the string and the null terminator, so a null terminated string will
> be decoded in any case, we cannot null terminate in pmi_decode unless an
> extra byte is calloc'ed
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/open-mpi/ompi/pull/1688#issuecomment-220982156>
>


-- 
-----
Best regards, Artem Polyakov
(Mobile mail)

ggouaillardet commented:
slurm uses ascii strings only, that is why binary is base64 encoded.
my question is why do we null terminate the decoded binary ?
binary is not sent to slurm, isn't it ?

artpol84 commented:
That is a good point. We shouldn't need null at the end and this is a bug
in my patch. I'll fix it now.

понедельник, 23 мая 2016 г. пользователь Gilles Gouaillardet написал:

> slurm uses ascii strings only, that is why binary is base64 encoded.
> my question is why do we null terminate the decoded binary ?
> binary is not sent to slurm, isn't it ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/open-mpi/ompi/pull/1688#issuecomment-220986718>
>


-- 
-----
Best regards, Artem Polyakov
(Mobile mail)

artpol84 commented:
This is a good point, @ggouaillardet. I fixed that. We either need to continue allocating +1 byte or do not add the '\0'.
If my understanding is correct, there is the only one datapath that uses pmi_decode:
* https://github.com/open-mpi/ompi/blob/master/opal/mca/pmix/base/pmix_base_fns.c#L509
* https://github.com/open-mpi/ompi/blob/master/opal/mca/pmix/base/pmix_base_fns.c#L553

And in the `for` loop here:
https://github.com/open-mpi/ompi/blob/master/opal/mca/pmix/base/pmix_base_fns.c#L553

in the exit condition we check for both `len` overflow and reaching of nul-terminator. For completeness I removed nul-terminator checking there.

artpol84 commented:
I've updated the branch but it seems that github experiences some problems so changes are not visible in the web interface yet:
https://twitter.com/githubstatus

however I can see them if I `git clone` my branch.

artpol84 commented:
:bot:retest

lanl-ompi commented:
Test FAILed.

ibm-ompi commented:
Build Failed with XL compiler! Please review the log, and get in touch if you have questions.

ibm-ompi commented:
Build Failed with GNU compiler! Please review the log, and get in touch if you have questions.

lanl-ompi commented:
Test FAILed.

lanl-ompi commented:
Test FAILed.

jjhursey commented:
Looking at the Jenkins from this it looks like it is related to the GitHub outage (the error is from doing a GitHub checkout)
```shell
 > git rev-parse b24546bf347e20600a1d0a6a893adf55978ac6aa^{commit} # timeout=10
ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job.
```

Give GitHub some time to resolve, and by try to trigger the bots to retest.

artpol84 commented:
Yes,
still not everything is correctly visible. Will need to give GH some time.

artpol84 commented:
For some reasons it shows @ggouaillardet ROMIO commit as this PR's commit even though it is already in OMPI master:
https://github.com/open-mpi/ompi/commit/b159587325aaf3a06aaba0b3f205b8c4fa032f26

Will wait for tomorrow and reopen if this won't be fixed.

lanl-ompi commented:
Test FAILed.

lanl-ompi commented:
Test FAILed.

ibm-ompi commented:
Build Failed with XL compiler! Please review the log, and get in touch if you have questions.

ibm-ompi commented:
Build Failed with GNU compiler! Please review the log, and get in touch if you have questions.

hppritcha commented:
I'm feeling really lucky.
bot:retest

rhc54 commented:
👍  tested on my Slurm setup under Slurm master, both PMI and PMI2.

jladd-mlnx commented:
bot:retest

hjelmn commented:
:bot:retest:

jladd-mlnx commented:
bot:retest

artpol84 commented:
what is the status? Can we merge?

artpol84 commented:
@ggouaillardet I plan to merge it if no objections.

ggouaillardet commented:
@artpol84 no objection, please go ahead !

artpol84 commented:
Thank you!
